Port wheel slip plugin to ros2 (eloquent)

scpeters opened the pull request:
This ports the wheel slip plugin from melodic (#995) to ros2. It uses ros2 parameters instead of dynamic reconfigure.

An example world is provided that shows two `trisphere_cycle` models that each have two instances of the wheel slip plugin, allowing separate parameters for the front and rear wheels. The demo world has some example `ros2 param` commands to adjusting the parameters dynamically.

scpeters committed:
gazebo_ros_wheel_slip: port skeleton to eloquent

scpeters committed:
Add parameter documentation to wheel slip plugin

scpeters committed:
Add parameter callback for slip compliance params

scpeters committed:
add demo world

scpeters committed:
add demo world instructions

scpeters commented:
I may need to rework this a bit to match the changes in https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1111

scpeters commented:
@jacobperron considering the change in #1126 in noetic that attempts to use the parameter value specified in the launch file, what would the equivalent be for ros2? I'm guessing this PR might need a change to match that behavior

scpeters commented:
I can make a new branch and target to foxy as well

chapulina commented:
> I can make a new branch and target to foxy as well

Note that technically Foxy has already been released (bloomed but not synced yet, still in [ros2-testing](http://packages.ros.org/ros2-testing/ubuntu/pool/main/r/ros-foxy-gazebo-plugins/)). So if you really want to break behaviour on Foxy, I'd recommend doing it soon. Otherwise, targeting Galactic.

scpeters commented:
> Note that technically Foxy has already been released (bloomed but not synced yet, still in ros2-testing). So if you really want to break behaviour on Foxy, I'd recommend doing it soon. Otherwise, targeting Galactic.

This is adding a new plugin, so I don't think it breaks anything

jacobperron commented on gazebo_plugins/src/gazebo_ros_wheel_slip.cpp:
```diff
@@ -0,0 +1,91 @@
+  impl_->ros_node_->declare_parameter("slip_compliance_unitless_lateral", 0.0);
+  impl_->ros_node_->declare_parameter("slip_compliance_unitless_longitudinal", 0.0);
+
+  auto existing_callback = impl_->ros_node_->set_on_parameters_set_callback(nullptr);
```
`set_on_parameters_set_callback` has been deprecated in Foxy. I recommend using [add_on_set_parameters_callback()](http://docs.ros2.org/eloquent/api/rclcpp/classrclcpp_1_1Node.html#a0c3ad2ca5c56eb54af0227f04e943679) as of Eloquent to avoid future deprecation warnings. This should also simplify the logic a little bit since we don't have to worry about handling any existing callbacks.

scpeters replied:
I've switched to this API in 00257a6

jacobperron commented on gazebo_plugins/src/gazebo_ros_wheel_slip.cpp:
```diff
@@ -0,0 +1,91 @@
+  impl_->ros_node_ = gazebo_ros::Node::Get(_sdf);
+
+  impl_->ros_node_->declare_parameter("slip_compliance_unitless_lateral", 0.0);
+  impl_->ros_node_->declare_parameter("slip_compliance_unitless_longitudinal", 0.0);
```
If you declare parameters *after* registering the parameters callback, then the callback will triggered when calling `declare_parameter`. Then if the defaults change in `declare_parameters`, the plugins values will also be updated.

scpeters replied:
ok, I changed the order of these function calls in 563c621

jacobperron commented:
> @jacobperron considering the change in #1126 in noetic that attempts to use the parameter value specified in the launch file, what would the equivalent be for ros2? I'm guessing this PR might need a change to match that behavior

I don't think we have the same problem in ROS 2, since native parameters are reconfigurable. A parameter passed to the node in a launch file should override the default value in `declare_parameter`.

scpeters committed:
Support parameter value override in launch file

Similar to the change in #1111, moves the WheelSlipPlugin::Load
call before the parameter callback so that the values from SDF/URDF
are set first.
Then the callback is changed to ignore negative values, and the
default slip parameter values are set to -1, so that the SDF/URDF
values are still preferred unless a different parameter value
is specified in a launch file.

scpeters commented:
I rebased and changed the behavior in 00257a609f0de78606dafe0671b852e69cf15f64 to match the change in #1111 / #1126.

scpeters committed:
fix cmake

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Call declare_parameters after adding callback

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
remove reference to existing_callback

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
fix lint errors

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

jacobperron commented on gazebo_plugins/src/gazebo_ros_wheel_slip.cpp:
```diff
@@ -0,0 +1,93 @@
+      return result;
+    };
+
+  impl_->ros_node_->add_on_set_parameters_callback(param_change_callback);
```
You might want to double-check, but I think we should maintain a reference to the returned callback handle, otherwise the callback may be unregistered.

scpeters replied:
yeah, it looks like this was done in https://github.com/ros2/rclcpp/pull/1136

scpeters replied:
I copied the approach from disparity_node in stereo_image_proc: https://github.com/ros-perception/image_pipeline/blob/2.2.0/stereo_image_proc/src/stereo_image_proc/disparity_node.cpp#L92-L182 in 0bef26d

scpeters committed:
Store handle to parameter set callback

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
fix build

scpeters commented:
@osrf-jenkins run tests please

scpeters commented:
@osrf-jenkins run tests please

nuclearsandwich commented:
@osrf-jenkins run tests please

scpeters commented:
some unrelated test failures, but CI is finally working
